File: src/geneticAlgorithm/systems/morphology/sequencer.cpp
#include "geneticAlgorithm/sequencer.hpp"
#include "modules/utils/genomeUtils.hpp"
#include <string>
#include <utility>
#include "modules/cuda/calculateAffinities.hpp"

void sequenceGRN(LifeForm& lifeForm) {
    // Temporary vars
    std::vector<Gene> factors;
    std::vector<Promoter> promoters;
    std::vector<Effector> effectors;
    std::vector<RegulatoryUnit> regulatoryUnits;

    // Temporary variables for regulatory units
    auto regulatoryUnit = RegulatoryUnit();
    std::vector<int> regulatoryPromoters;
    std::vector<int> regulatoryFactors;

    bool readingPromoters = true;
    for (auto [id, sequence] : lifeForm.genome.hoxGenes) {
        std::string rna = std::move(sequence); // Create rna copy
        try {
            int type = (int)(readUniqueBaseRange(rna, 2) * 16) % 5;
            bool sign = readBase(rna) >= 2;
            bool active = readBase(rna) >= 1;
            if (!active) continue;

            // Uniformly distributed modifiers (also allows for large jumps genetically)
            float modifier = readUniqueBaseRange(rna, 8);

            float3 embedding = make_float3(
                readUniqueBaseRange(rna, 8),
                readUniqueBaseRange(rna, 8),
                readUniqueBaseRange(rna, 8));

            // External factors (inputs to grn)
            if (type == 0) {
                Gene::FactorType externalFactorTypes[] = {
                    Gene::FactorType::MaternalFactor,
                    Gene::FactorType::Crowding,
                    Gene::FactorType::Constant,
                    Gene::FactorType::Generation,
                    Gene::FactorType::Energy,
                    Gene::FactorType::Time,
                };
                int subType = (int)(readUniqueBaseRange(rna, 3) * 64) % 6;
                Gene::FactorType externalFactorType = externalFactorTypes[subType];

                float2 extra = {readUniqueBaseRange(rna, 8), readUniqueBaseRange(rna, 8)};
                Gene externalFactor = Gene(externalFactorType,
                                           sign, modifier, embedding, extra);
                factors.push_back(externalFactor);
            }

            // Gene effectors (outputs of grn)
            if (type == 1) {
                Effector::EffectorType subType = static_cast<Effector::EffectorType>(
                    (int)(readUniqueBaseRange(rna, 4) * 256) % (int)Effector::EffectorType::EFFECTOR_LENGTH);

                // If this grn already contains a special node
                // (e.g. maternal factor, crowding, etc.) of this type, skip
                if (std::any_of(effectors.begin(), effectors.end(),
                                [subType](const Effector& effector) {
                                    return effector.effectorType == subType;
                                })) {
                    continue;
                }

                Effector effector = Effector(subType,
                                             sign, modifier, embedding);
                effectors.push_back(effector);
            }

            // Regulatory units (nodes in grn)
            // Promoters (take in factors and produce activity levels excitatory/inhibitory additive/multiplicative)
            if (type == 2) {
                int additive = readBase(rna) >= 1;
                Promoter::PromoterType promoterType = additive
                                                          ? Promoter::PromoterType::Additive
                                                          : Promoter::PromoterType::Multiplicative;
                auto promoter = Promoter(promoterType, sign, modifier, embedding);

                if (!readingPromoters) {
                    regulatoryUnit.promoters = StaticGPUVector(regulatoryPromoters);
                    regulatoryUnit.factors = StaticGPUVector(regulatoryFactors);
                    regulatoryUnits.push_back(regulatoryUnit);
                    regulatoryUnit = RegulatoryUnit();
                    readingPromoters = true;
                }

                regulatoryPromoters.push_back(promoters.size());
                promoters.push_back(promoter);
            }

            // Genes - internal product, external product, receptor
            if (type == 3 || type == 4 || type == 5) {
                Gene::FactorType geneTypes[] = {
                    Gene::FactorType::ExternalProduct,
                    Gene::FactorType::InternalProduct,
                    Gene::FactorType::Receptor,
                };

                Gene gene = Gene(geneTypes[type - 3], sign, modifier, embedding);

                if (regulatoryPromoters.empty()) continue;
                readingPromoters = false;
                regulatoryFactors.push_back((int)factors.size());
                factors.push_back(gene);
            }
        }
        catch (RNAExhaustedException& e) {
            // "RNA exhausted"
        }
    }

    lifeForm.grn.factors = StaticGPUVector(factors);
    lifeForm.grn.promoters = StaticGPUVector(promoters);
    lifeForm.grn.effectors = StaticGPUVector(effectors);

    lifeForm.grn.regulatoryUnits = StaticGPUVector(regulatoryUnits);
    calculateAffinities(lifeForm.grn);
}

File: rust/src/genetic_algorithm/sequencer.rs
// Sequencer - reads genome and builds gene regulatory network

use std::collections::VecDeque;

use crate::genetic_algorithm::genome::Genome;
use crate::genetic_algorithm::systems::{GeneRegulatoryNetwork, Receptor, Factor, Promoter, Effector,
     RegulatoryUnit, ReceptorType, FactorType, PromoterType, EffectorType, EMBEDDING_DIMENSIONS};
use crate::genetic_algorithm::utils::{read_base, read_base_range, read_unique_base_range};

/// Sequence a genome to build a gene regulatory network
/// Returns the GRN even if some genes fail to parse
pub fn sequence_grn(genome: &Genome) -> GeneRegulatoryNetwork {
    let mut grn = GeneRegulatoryNetwork::new();
    // Inputs
    let mut receptors: Vec<Receptor> = Vec::new();
    // hidden units
    let mut regulatory_units: Vec<RegulatoryUnit> = Vec::new();
    // Outputs
    let mut effectors: Vec<Effector> = Vec::new();
    
    let mut regulatory_unit = RegulatoryUnit::new();
    let mut regulatory_promoters: Vec<Promoter> = Vec::new();
    let mut regulatory_factors: Vec<Factor> = Vec::new();
    let mut reading_promoters = true;
    
    // Process each gene in the genome
    for (_gene_id, sequence) in genome
        .hox_gene_order
        .iter()
        .filter_map(|id| genome.hox_genes.get(id).map(|seq| (*id, seq.clone())))
    {
        let mut rna: VecDeque<u8> = VecDeque::from(sequence.clone());
        
        // Process genes until RNA is exhausted or we can't read more
        while rna.len() >= 20 { // Minimum required bases for a gene (2 + 1 + 1 + 8 + 8*3 = 20)
            // Read gene type (2 bases -> 0-16, mod 5 -> 0-4)
            let type_val = match read_base_range(&mut rna, 2) {
                Ok(val) => (val * 16.0) as usize % 5,
                Err(_) => break,
            };
            let sign = match read_base(&mut rna) {
                Ok(b) => b >= 2,
                Err(_) => break,
            };
            let active = match read_base(&mut rna) {
                Ok(b) => b >= 1,
                Err(_) => break,
            };
            
            if !active {
                continue;
            }
            
            // Read modifier (8 bases)
            let modifier = match read_unique_base_range(&mut rna, 8) {
                Ok(m) => m,
                Err(_) => break,
            };
            
            // Read embedding (8 bases each for x, y, z ... of EMBEDDING_DIMENSIONS)
            let mut embedding: [f32; EMBEDDING_DIMENSIONS] = [0.0f32; EMBEDDING_DIMENSIONS];
            for i in 0..EMBEDDING_DIMENSIONS {
                embedding[i] = match read_unique_base_range(&mut rna, 8) {
                    Ok(m) => m,
                    Err(_) => break,
                };
            }
            
            // Process based on type
            match type_val {
                0 => {
                    // Receptor
                    let receptor_types = [
                        ReceptorType::MaternalFactor,
                        ReceptorType::Crowding,
                        ReceptorType::Constant,
                        ReceptorType::Generation,
                        ReceptorType::Energy,
                        ReceptorType::Time,
                    ];
                    let sub_type = match read_unique_base_range(&mut rna, 3) {
                        Ok(val) => ((val * 64.0) as usize) % 6,
                        Err(_) => break,
                    };
                    let receptor_type: ReceptorType = receptor_types[sub_type];
                    
                    let extra = match (
                        read_unique_base_range(&mut rna, 8),
                        read_unique_base_range(&mut rna, 8),
                    ) {
                        (Ok(x), Ok(y)) => [x, y],
                        _ => break,
                    };
                    
                    let receptor = Receptor::new(receptor_type, sign, modifier, embedding, extra);
                    receptors.push(receptor);
                }
                1 => {
                    // Effectors
                    let effector_types = [
                        EffectorType::Divide, EffectorType::Die, EffectorType::Freeze,
                        EffectorType::Distance, EffectorType::Radius,
                        EffectorType::Red, EffectorType::Green, EffectorType::Blue,
                    ];
                    let sub_type = match read_unique_base_range(&mut rna, 4) {
                        Ok(val) => ((val * 256.0) as usize) % effector_types.len(),
                        Err(_) => break,
                    };
                    let effector_type = effector_types[sub_type];

                    let effector = Effector::new(effector_type, sign, modifier, embedding);
                    effectors.push(effector);
                }
                2 => {
                    // Promoters
                    let additive = match read_base(&mut rna) {
                        Ok(b) => b >= 1,
                        Err(_) => break,
                    };
                    let promoter_type = if additive {
                        PromoterType::Additive
                    } else {
                        PromoterType::Multiplicative
                    };
                    
                    let promoter = Promoter::new(promoter_type, sign, modifier, embedding);
                    
                    if !reading_promoters {
                        // Finish current regulatory unit
                        regulatory_unit.promoters = regulatory_promoters.clone();
                        regulatory_unit.factors = regulatory_factors.clone();
                        regulatory_units.push(regulatory_unit);
                        regulatory_unit = RegulatoryUnit::new();
                        regulatory_promoters.clear();
                        regulatory_factors.clear();
                        reading_promoters = true;
                    }
                    
                    regulatory_promoters.push(promoter);
                }
                3 | 4 | 5 => {
                    // Genes: internal product, external product, receptor
                    let gene_types = [
                        FactorType::ExternalMorphogen,
                        FactorType::InternalMorphogen,
                        FactorType::Orientant,
                    ];
                    
                    if regulatory_promoters.is_empty() {
                        break;
                    }
                    
                    let factor_type = gene_types[type_val - 3];
                    let gene = Factor::new(factor_type, sign, modifier, embedding);
                    
                    reading_promoters = false;
                    regulatory_factors.push(gene);
                }
                _ => {}
            }
        }
    }
    
    // Add final regulatory unit if it exists
    if !regulatory_promoters.is_empty() && !regulatory_factors.is_empty() {
        regulatory_unit.promoters = regulatory_promoters;
        regulatory_unit.factors = regulatory_factors;
        regulatory_units.push(regulatory_unit);
    }
    
    grn.receptors = receptors;
    grn.effectors = effectors;
    grn.regulatory_units = regulatory_units;
    grn
}

